crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+        let lazy_load_members = true;
+
+        let response = self.room.event_with_context(event_id, lazy_load_members).await?;
+
+        let has_prev = response.prev_batch_token.is_some();
+        let has_next = response.next_batch_token.is_some();
+        *self.prev_batch_token.lock().await = response.prev_batch_token;
+        *self.next_batch_token.lock().await = response.next_batch_token;
+
+        self.state.set(PaginatorState::Idle);
+
+        // Consolidate the events into a linear timeline, topologically ordered.
+        // - the events before are returned in the reverse topological order: invert
+        //   them.
+        // - insert the target event, if set.
+        // - the events after are returned in the correct topological order.
+
+        let events = response
+            .events_before
+            .into_iter()
+            .rev()
+            .chain(response.event)
+            .chain(response.events_after)
+            .collect();
+
+        Ok(StartFromResult { events, has_prev, has_next })
+    }
+
+    /// Runs a backward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the start of the timeline before.
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_backward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Backward).await
+    }
+
+    /// Runs a forward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the end of the timeline before.
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_forward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Forward).await
+    }
+
+    async fn paginate(&self, dir: Direction) -> Result<PaginationResult, PaginatorError> {
+        self.check_state(PaginatorState::Idle)?;
+
+        let token = match dir {
+            Direction::Backward => {
+                let prev_batch_token = self.prev_batch_token.lock().await;
+                if prev_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                prev_batch_token.clone()
+            }
+
+            Direction::Forward => {
+                let next_batch_token = self.next_batch_token.lock().await;
+                if next_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                next_batch_token.clone()
+            }
+        };
+
+        // Note: it's possible two callers have checked the state and both figured it's
+        // idle. This check makes sure there's at most one which can set the
+        // state to paginating, preventing a race condition.
```

Comment:
Are these kinds of errors expected to happen? If so, how will we handle them?

Reply:
Also, should the `check_state` call be as close as possible to `set_if_not_eq`?

Reply:
> Are these kinds of errors expected to happen? If so, how will we handle them?

I don't understand these two questions. There are no errors that can happen, for the reasons explained in the code comment.

> Also, should the check_state call be as close as possible to set_if_not_eq?

No, because it'd let one believe that `paginate_backward()` works and returns nothing before calling `start_from()` (since the `prev_batch_token` would be none, and we'd short-circuit the return).

Reply:
> No, because it'd let one believe that paginate_backward() works and returns nothing before calling start_from() (since the prev_batch_token would be none, and we'd short-circuit the return).

I guess we could get the state to `Paginating` early, and unset it during the early returns, but it's probably not worth it. I'm still wondering whether we think this race where we check the state and then it changes while we are running this code is going to happen. If so, the early return might be misleading, right?

Reply:
> I mean the Err(PaginatorError::InvalidPreviousState that we return. Is this expected to happen?

Spawn two tasks, each calling the method, while the initial state is `Idle`? It's a race condition, and it would likely require total control over tokio to trigger…

> If so, the early return might be misleading, right?

I don't think it is: having the initial state be "idle" is an invariant, so should be checked at all times; then having two callers race on it is rather exceptional, and what the second check is preventing.